File: src/data_segregation_lab/tool_protocol.py
"""Fail-closed parser and sanitizer for the lab's intentionally tiny text tool protocol."""

from __future__ import annotations

import re
from typing import cast

from data_segregation_lab.models import ToolAction, ToolCall

_KEY_PATTERN = re.compile(r"^[A-Za-z0-9_.-]{1,64}$")
_MAX_VALUE_CHARS = 256


class ToolCallParseError(ValueError):
    """Raised internally when a tool line is ambiguous or incomplete."""


def _parse_arguments(parts: list[str]) -> dict[str, str]:
    """Parse unquoted ``name=value`` tokens and reject duplicate names."""
    arguments: dict[str, str] = {}
    for part in parts:
        if "=" not in part:
            raise ToolCallParseError(f"Missing '=' in tool argument: {part}")
        name, value = part.split("=", 1)
        if not name or not value:
            raise ToolCallParseError("Tool argument names and values cannot be empty")
        if name in arguments:
            raise ToolCallParseError(f"Duplicate tool argument: {name}")
        if value.startswith('"') and value.endswith('"'):
            value = value[1:-1]
        arguments[name] = value
    return arguments


def _parse_tool_line(line: str) -> ToolCall:
    """Parse one normalized ``[tool]`` line with action-specific fields."""
    parts = line.split()
    if len(parts) < 4:
        raise ToolCallParseError("Tool call is incomplete")

    raw_action = parts[1]
    if raw_action not in {"read", "write"}:
        raise ToolCallParseError(f"Unsupported tool action: {raw_action}")
    action = cast(ToolAction, raw_action)
    arguments = _parse_arguments(parts[2:])

    expected = {"owner", "key", "value"} if action == "write" else {"owner", "key"}
    if arguments.keys() != expected:
        raise ToolCallParseError(
            f"{action} requires exactly these arguments: {sorted(expected)}"
        )

    return ToolCall(
        action=action,
        owner=arguments["owner"],
        key=arguments["key"],
        value=arguments.get("value"),
    )


def detect_tool_calls(text: str) -> list[ToolCall]:
    """Extract valid tool calls while ignoring prose and malformed proposals."""
    calls: list[ToolCall] = []
    for raw_line in text.splitlines():
        line = raw_line.strip()
        if line.startswith("EXEC:"):
            line = line.removeprefix("EXEC:").strip()
        if not line.startswith("[tool]"):
            continue
        try:
            calls.append(_parse_tool_line(line))
        except ToolCallParseError:
            # Model text is untrusted. Invalid operations never reach storage.
            continue
    return calls


def sanitize_tool_call(call: ToolCall, allowed_owners: frozenset[str]) -> ToolCall | None:
    """Validate a parsed tool call before it reaches the authorization boundary.

    Ported from the ``sanitizeDecision`` pattern in Capability Wall: even a
    hijacked model cannot name an unknown tenant, use an malformed key, or ship
    an unbounded write payload.
    """
    owner = call.owner.strip()
    if owner not in allowed_owners:
        return None

    key = call.key.strip()
    if not _KEY_PATTERN.fullmatch(key):
        return None

    if call.action == "write":
        value = (call.value or "").strip()
        if len(value) > _MAX_VALUE_CHARS:
            value = value[:_MAX_VALUE_CHARS]
        return ToolCall(call.action, owner, key, value)

    return ToolCall(call.action, owner, key, None)

File: shared/toolProtocol.ts
import type { ToolCall } from "./models.js";
import { PARTICIPANTS } from "./protocol.js";

const KEY_PATTERN = /^[A-Za-z0-9_.-]{1,64}$/;
const MAX_VALUE_CHARS = 256;

class ToolCallParseError extends Error {}

function parseArguments(parts: string[]): Record<string, string> {
  const arguments_: Record<string, string> = {};
  for (const part of parts) {
    if (!part.includes("=")) {
      throw new ToolCallParseError(`Missing '=' in tool argument: ${part}`);
    }
    const eq = part.indexOf("=");
    const name = part.slice(0, eq);
    let value = part.slice(eq + 1);
    if (!name || !value) {
      throw new ToolCallParseError("Tool argument names and values cannot be empty");
    }
    if (name in arguments_) {
      throw new ToolCallParseError(`Duplicate tool argument: ${name}`);
    }
    if (value.startsWith('"') && value.endsWith('"')) {
      value = value.slice(1, -1);
    }
    arguments_[name] = value;
  }
  return arguments_;
}

function parseToolLine(line: string): ToolCall {
  const parts = line.split(/\s+/);
  if (parts.length < 4) throw new ToolCallParseError("Tool call is incomplete");

  const rawAction = parts[1];
  if (rawAction !== "read" && rawAction !== "write") {
    throw new ToolCallParseError(`Unsupported tool action: ${rawAction}`);
  }

  const arguments_ = parseArguments(parts.slice(2));
  const expected =
    rawAction === "write"
      ? new Set(["owner", "key", "value"])
      : new Set(["owner", "key"]);

  const keys = new Set(Object.keys(arguments_));
  if (keys.size !== expected.size || ![...expected].every((k) => keys.has(k))) {
    throw new ToolCallParseError(
      `${rawAction} requires exactly these arguments: ${[...expected].sort().join(", ")}`,
    );
  }

  return {
    action: rawAction,
    owner: arguments_.owner,
    key: arguments_.key,
    value: rawAction === "write" ? arguments_.value : null,
  };
}

export function detectToolCalls(text: string): ToolCall[] {
  const calls: ToolCall[] = [];
  for (const rawLine of text.split("\n")) {
    let line = rawLine.trim();
    if (line.startsWith("EXEC:")) line = line.slice(5).trim();
    if (!line.startsWith("[tool]")) continue;
    try {
      calls.push(parseToolLine(line));
    } catch {
      continue;
    }
  }
  return calls;
}

export function sanitizeToolCall(call: ToolCall): ToolCall | null {
  const owner = call.owner.trim();
  if (!PARTICIPANTS.has(owner)) return null;

  const key = call.key.trim();
  if (!KEY_PATTERN.test(key)) return null;

  if (call.action === "write") {
    let value = (call.value ?? "").trim();
    if (value.length > MAX_VALUE_CHARS) value = value.slice(0, MAX_VALUE_CHARS);
    return { action: call.action, owner, key, value };
  }

  return { action: call.action, owner, key, value: null };
}

export function formatToolCall(call: ToolCall): string {
  let fields = `owner="${call.owner}", key="${call.key}"`;
  if (call.value != null) fields += `, value="${call.value}"`;
  return `${call.action}(${fields})`;
}
